productivity-suite/limitations-and-improvements.md

```diff
@@ -0,0 +1,63 @@
+# Limitations and Areas of Improvements
+
+In this document we discuss various limitations and are of improvements either related to the Productivity Suite implementation or the piercing strategy as a whole.
+
+## Pre Piercing Positioning
+
+One of the basic challenges provided by the strategy consists in making the piercing operation unobtrusive and not jarring for the user.
+
+In order to do so the fragments present before the application bootstraps (the ones included in the initial server-side rendered stream) need to be positioned as close as possible to their final location inside the legacy application.
+
+We’ve accomplished this by setting pre-piercing styles to the fragments which absolutely position the fragments in the appropriate locations.
+
+You can see the `prePiercingStyles` present in the `productivity-suite/app/legacy-app/src/worker/script.ts` file.
+
+What these styles do is to place the fragments in the right position before the application bootstraps, and, since there are different media queries and dynamic contents in the legacy application, media queries needed to be added to these styles so that the fragments' vertical alignment can be accordingly adjusted.
+
+This method is very manual and error prone, it doesn’t work well with different browser font-sizes and it is unlikely to produce pixel-perfect results for any screen resolution and window size.
+
+An alternative solution that we’ve thought of consists in instead providing a means to add some scaffolding/placeholder html around the fragment so that such html can be styled to mimic the legacy app’s elements and their styles.
+
+Such a method could provide a more robust positioning solution but it would also introduce new challenges which would need to be addressed, such as dx related concerns (e.g. how can the html be added in a simple way which doesn’t promote code duplication?) and issues related to dynamic content present in the legacy application (e.g. how can the scaffolding elements reflect the side of elements in the legacy application which sizes derive from dynamic content).
+Although not fail proof this is a different avenue for the fragments positioning which would definitely be worth exploring and which could provide a nicer and more robust solution to the positioning issue.
+
+## Side-Effectful Module Type Scripts
```

Comment:
mh... I think it's fine, I always used that Module Type Script sort of definition to make a more clear distinction of the issue since you can have for example:
 `<script type="module">console.log('hello world');</script>`
in which ES module imports/exports are nowhere to be found (but we'd still have the same issue here)

The browser does create an ES module regardless...

Do you think that mentioning directly ES modules could make things less clear?

Reply:
I am not worried about adding the ES bit. But I found that I wasn't sure what "Module Type Script" meant on first reading. Not a blocker.

Reply:
in the HTML specs they call them [Javascript Module Script](https://html.spec.whatwg.org/multipage/webappapis.html#javascript-module-script)s, so what would you think about:
```suggestion
## Javascript module scripts with side-effects
```
?
I think it is clear and accurate 😄

Reply:
Just `Module scripts with side-effects` would be enough for me!